python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -214,79 +242,83 @@ def __init__(
-            # other's node ID. We cannot use a separate concurrency group to execute the
-            # function because reader actors may not have been declared with an
-            # additional concurrency group beyond default.
-            #
-            # TODO(jhumphri): Allow different readers for the same channel to be on
-            # different nodes.
-            prev_reader_node = None
-            prev_reader = None
-            for reader, node in reader_and_node_list:
-                if prev_reader_node is None:
-                    prev_reader_node = node
-                elif prev_reader_node != node:
-                    raise ValueError(
-                        f"All reader actors must be on the same node. Actor "
-                        f"{prev_reader} is on node {prev_reader_node} while actor "
-                        f"{reader} is on node {node}."
-                    )
-                prev_reader = reader
+            for reader, node_id in reader_and_node_list:
+                self._reader_node_ids.add(node_id)
 
             self._writer_node_id = (
                 ray.runtime_context.get_runtime_context().get_node_id()
             )
             self._writer_ref = _create_channel_ref(self, typ.buffer_size_bytes)
-            self._reader_node_id = prev_reader_node
-
-            self._create_reader_ref(reader_and_node_list, typ.buffer_size_bytes)
 
-            assert self._reader_ref is not None
+            self._create_reader_refs(typ.buffer_size_bytes)
         else:
             assert (
                 _writer_node_id is not None
             ), "_writer_node_id must also be passed to the constructor when "
             "_writer_ref is."
-            assert (
-                _reader_ref is not None
-            ), "_reader_ref must also be passed to the constructor when _writer_ref is."
+            assert _node_id_to_reader_info is not None, (
+                "_node_id_to_reader_info must also be passed to the constructor "
+                "when _writer_ref is."
+            )
 
             self._writer_ref = _writer_ref
             self._writer_node_id = _writer_node_id
-            self._reader_node_id = _reader_node_id
-            self._reader_ref = _reader_ref
-
-        self._num_readers = len(self._reader_and_node_list)
-        if self.is_remote():
-            # Even though there may be multiple readers on a remote node, we set
-            # `self._num_readers` to 1 here. On this local node, only the IO thread in
-            # the mutable object provider will read the mutable object. The IO thread
-            # will then send a gRPC with the mutable object contents to the remote node
-            # where the readers are.
-            self._num_readers = 1
-
-    def _create_reader_ref(
+            self._reader_node_ids = _reader_node_ids
+            self._node_id_to_reader_info = _node_id_to_reader_info
+
+        assert self._num_local_readers == 0
```

Comment:
This is a bit weird, it is set to 0 at L233, and asserted here. Why not just set it here?